pulsar-client/src/main/java/org/apache/pulsar/client/impl/transaction/TransactionImpl.java

```diff
@@ -269,11 +250,18 @@ public TxnID getTxnID() {
                 + state.name() + ", expect " + State.OPEN + " state!"));
     }
 
+    public synchronized void registerSendOp(CompletableFuture<MessageId> newSendFuture) {
+        sendFuture = sendFuture.thenCompose(ignore -> newSendFuture);
+    }
+
+    public synchronized void registerAckOp(CompletableFuture<Void> newAckFuture) {
+        ackFuture = ackFuture.thenCompose(ignore -> newAckFuture);
+    }
 
-    private CompletableFuture<Void> allOpComplete() {
+    public CompletableFuture<Void> prepareCommit() {
```

Comment:
The futures for sending messages, acking message should be handled by users, they are using the FutureUtil or any other tools to process the returned future.

Introduce the prepareCommit looks like we introduced a new concept about Pulsar transactions, but essentially it's just a tool for asynchronous processing

Reply:
This is really an important point, thanks a lot for the reminder. I have made optimizations, what do you think about this change?